src/machine/core.cpp

```diff
@@ -324,6 +351,14 @@ DecodeState Core::decode(const FetchInterstage &dt) {
     ExceptionCause excause = dt.excause;
 
     dt.inst.flags_alu_op_mem_ctl(flags, alu_op, mem_ctl);
+    CSR::PrivilegeLevel inst_xret_priv = CSR::PrivilegeLevel::UNPRIVILEGED;
+    if (flags & IMF_XRET) {
+        inst_xret_priv = decode_xret_type_from_inst(dt.inst);
+        // Mark illegal if current privilege is lower than encoded xRET type (e.g. MRET executed in S-mode)
+        if (state.current_privilege() < inst_xret_priv) {
+            excause = EXCAUSE_INSN_ILLEGAL;
+        }
```

Comment:
This change should not be needed. Should be solved by
```
const InstructionFlags check_inst_flags_val;
const InstructionFlags check_inst_flags_mask;
```
manipulation in `set_current_privilege` and snntation of the instructions by required mode `IMF_PRIV_S`, `IMF_PRIV_H`, `IMF_PRIV_M` in decoding tables.